user/api.go

```diff
@@ -714,53 +651,78 @@ func (a *Api) sendError(res http.ResponseWriter, statusCode int, reason string,
 	sendModelAsResWithStatus(res, status.NewStatus(statusCode, reason), statusCode)
 }
 
-func (a *Api) authenticateSessionToken(sessionToken string) (*TokenData, error) {
-	if sessionToken == "" {
+func (a *Api) authenticateToken(r *http.Request) (*TokenData, error) {
+	if sessionToken := r.Header.Get(TP_SESSION_TOKEN); sessionToken != "" {
+		return a.authenticateSessionToken(sessionToken)
+	}
+	return a.accessTokenChecker.Check(r)
+}
+
+func (a *Api) authenticateSessionToken(token string) (*TokenData, error) {
+	if token == "" {
 		return nil, errors.New("Session token is empty")
-	} else if tokenData, err := UnpackSessionTokenAndVerify(sessionToken, a.ApiConfig.Secret); err != nil {
+	}
+	tokenData, err := UnpackSessionTokenAndVerify(token, a.ApiConfig.Secret)
+	if err != nil {
+		if r, rErr := http.NewRequest("GET", "/", nil); rErr == nil {
```

Comment:
According to the `UnpackSessionTokenAndVerify` function, it can return an error if the token is a Tidepool Session Token, but it is expired. I don't think we want to query Auth0 in this case.

Also, do we really need to create an entirely new request to do this? Looking at the `validator.Claims(r, token, &claims)` call in the `AccessTokenChecker` code above, it calls `validator := auth0.NewValidator(configuration)`.  Rather that using the default validator, I think we can create a new `JWTValidator` by hand and specify our own `RequestTokenExtractor` which looks first in the `Authorization` header and then in the `X-Tidepool-Session-Token` header.